src/main/java/at/ac/tuwien/kr/alpha/common/atoms/BuiltinAtom.java

```diff
@@ -0,0 +1,222 @@
+			throw new RuntimeException("Builtin equality with left and right side being variables encountered. Should not happen.");
+		}
+		if (isLeftAssigning()) {
+			return Collections.singletonList((VariableTerm)terms.get(0));
+		}
+		if (isRightAssigning()) {
+			return Collections.singletonList((VariableTerm)terms.get(1));
+		}
+		return Collections.emptyList();
+	}
+
+	@Override
+	public List<VariableTerm> getNonBindingVariables() {
+		Term left = terms.get(0);
+		Term right = terms.get(1);
+		HashSet<VariableTerm> occurringVariables = new HashSet<>();
+		List<VariableTerm> leftOccurringVariables = new LinkedList<>(left.getOccurringVariables());
+		List<VariableTerm> rightOccurringVariables = new LinkedList<>(right.getOccurringVariables());
+		if (isLeftAssigning()) {
+			leftOccurringVariables.remove(left);
+		}
+		if (isRightAssigning()) {
+			rightOccurringVariables.remove(right);
+		}
+		occurringVariables.addAll(leftOccurringVariables);
+		occurringVariables.addAll(rightOccurringVariables);
+		if (isLeftAssigning() || isRightAssigning()) {
+			return new ArrayList<>(occurringVariables);
+		}
+		// Neither left- nor right-assigning, hence no variable is binding.
+		return new ArrayList<>(occurringVariables);
+	}
+
+	@Override
+	public Atom substitute(Substitution substitution) {
+		return new BuiltinAtom(terms.get(0).substitute(substitution),
+			terms.get(1).substitute(substitution),
+			negated, operator);
+	}
+
+	@Override
+	public List<Substitution> getSubstitutions(Substitution partialSubstitution) {
+		// Treat case that this is X = t or t = X.
+		if (isLeftAssigning() || isRightAssigning()) {
+			VariableTerm variable = null;
+			Term expression = null;
+			if (isLeftAssigning()) {
+				variable = (VariableTerm) terms.get(0);
+				expression = terms.get(1);
+			}
+			if (isRightAssigning()) {
+				variable = (VariableTerm) terms.get(1);
+				expression = terms.get(0);
+			}
+			Term groundTerm = expression.substitute(partialSubstitution);
+			Term resultTerm = null;
+			// Check if the groundTerm is an arithmetic expression and evaluate it if so.
+			if (groundTerm instanceof ArithmeticTerm) {
+				Integer result = ArithmeticTerm.evaluateGroundTerm(groundTerm);
+				if (result == null) {
+					return Collections.emptyList();
+				}
+				resultTerm = ConstantTerm.getInstance(result);
+			} else {
+				// Ground term is another term (constant, or function term).
+				resultTerm = groundTerm;
+			}
+			Substitution extendedSubstitution = new Substitution(partialSubstitution);
+			extendedSubstitution.put(variable, resultTerm);
+			return Collections.singletonList(extendedSubstitution);
+		}
+
+		// Treat all other cases, i.e., this is just comparison with all variables bound by partialSubstitution.
+		if (compare(terms.get(0).substitute(partialSubstitution), terms.get(1).substitute(partialSubstitution))) {
```

Comment:
Just a matter of style, so you might as well disregard this comment. I like to have they "easy"/"simple" cases of a function at the top. Fox example, if you would move these two cases to the top and wrap them inside an `if (!isLeftAssigning() && !isRightAssigning())` the whole block above (which is more "complicated") would have one less indentation.

Reply:
Good point. Thanks!